src/Microsoft.AspNet.Mvc.Localization/HtmlLocalizer.cs

```diff
@@ -0,0 +1,96 @@
+    {
+        private readonly IStringLocalizer _localizer;
+        private readonly IHtmlEncoder _encoder;
+
+        /// <summary>
+        /// Creates a new <see cref="HtmlLocalizer"/>.
+        /// </summary>
+        /// <param name="localizer">The <see cref="IStringLocalizer"/> to read strings from.</param>
+        /// <param name="encoder">The <see cref="IHtmlEncoder"/>.</param>
+        public HtmlLocalizer([NotNull] IStringLocalizer localizer, [NotNull] IHtmlEncoder encoder)
+        {
+            _localizer = localizer;
+            _encoder = encoder;
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="HtmlLocalizer"/> for a specific <see cref="CultureInfo"/>.
+        /// </summary>
+        /// <param name="culture">The <see cref="CultureInfo"/> to use.</param>
+        /// <returns>A culture-specific <see cref="HtmlLocalizer"/>.</returns>
+        public virtual IHtmlLocalizer WithCulture([NotNull] CultureInfo culture) =>
+            new HtmlLocalizer(_localizer.WithCulture(culture), _encoder);
+
+        IStringLocalizer IStringLocalizer.WithCulture([NotNull] CultureInfo culture) => WithCulture(culture);
+
+        /// <inheritdoc />
+        public virtual LocalizedString this[string key] => _localizer[key];
+
+        /// <inheritdoc />
+        public virtual LocalizedString this[string key, params object[] arguments] => _localizer[key, arguments];
+
+        /// <inheritdoc />
+        public virtual LocalizedString GetString([NotNull] string key) => _localizer.GetString(key);
+
+        /// <inheritdoc />
+        public virtual LocalizedString GetString([NotNull] string key, params object[] arguments) =>
+            _localizer.GetString(key, arguments);
+
+        /// <inheritdoc />
+        public IEnumerable<LocalizedString> GetAllStrings(bool includeAncestorCultures)=>
+            _localizer.GetAllStrings(includeAncestorCultures);
+
+        /// <inheritdoc />
+        public virtual LocalizedHtmlString Html([NotNull] string key) => ToHtmlString(_localizer.GetString(key));
+
+        /// <inheritdoc />
+        public virtual LocalizedHtmlString Html([NotNull] string key, params object[] arguments)
+        {
+            return ToHtmlString(_localizer.GetString(key, EncodeArguments(arguments)));
+        }
+
+        protected LocalizedHtmlString ToHtmlString(LocalizedString result)
+        {
+            return new LocalizedHtmlString(result.Name, result.Value, result.ResourceNotFound);
+        }
+
+        protected object[] EncodeArguments(object[] arguments)
+        {
+            object[] encodedArguments = new object[arguments.Length];
+            for (var index = 0; index != arguments.Length; ++index)
+            {
+                var argument = arguments[index];
+                if (argument.GetType().GetTypeInfo().IsPrimitive ||
+                    argument is HtmlString ||
+                    argument is DateTime ||
+                    argument is DateTimeOffset)
```

Comment:
FYI the general rules for HTML / JavaScript / URL (mostly HTML) encoding are:
1. Trust HTML strings the user placed in the .cshtml file ... but verify their typing, rejecting anything Razor can't parse.
2. Trust HTML strings HTML and tag helpers provide .. but not every string e.g. `@Html.Value()` returns a `string` that will be encoded.
3. Don't trust anything else.

The implementation of the above is fairly simple: `HtmlString`, `TagHelperContent`, and a few other wrappers contain strings that don't need to be encoded. Everything else _must_ be encoded. Likely this code only needs to deal with `HtmlString`.